### ISSUE TITLE
Cancelling node move operations

### ISSUE THREAD

**Issue by rrhvella** [fritz-c/react-sortable-tree#122]
Type: Feature Request

I am using this component to reflect a backend tree structure, and we would like to allow the user to perform re-parenting and re-ordering commands which are immediately sent to the server once they are performed. Since these commands are sent to the server, it means that they can fail. 

In order to fulfill this requirement, I need detailed information of the operation performed in the onChange event so that I can.

A: Send the request to the server based on the operation performed.
B: Selectively accept or reject the change based on the response from the server.

At least, this is the most elegant way I can think of to solve this problem. Obviously, I can compare the new tree to the old tree in order to derive the operation performed, but I think it would be much easier on the whole if I got that information in the onChange event. 

Is this possible or is there an easier way to fullfill this requirement through the existing api?

Thank you for the excellent component.

**Comment by rrhvella** [comment]
After investigating the source code a bit more, I realized that the same affect can be achieved by hooking to the more granular transformation events (onVisibilityToggle and onMoveNode) and ignoring the onChange event. 

There are a couple of problems with this though:
* Operations such as expansion due to lazy loading only fire an onChange event, they do not fire a more granular transformation event.
* onChange is required, and hooking to both onChange and the more granular transformation events would be incorrect.

I can work with things as they are, but if people like this idea, I can perform a pull request. I would like people's input on weather this is a good idea and which approach they prefer before doing any work. I think in general these features would be useful.

**Comment by rrhvella** [comment]
I should also point out that since the state is updated based on the data prop, rather than being a direct reflection of the data prop, that in order to cancel an operation, I have to do a shallow copy of the existing tree. I do not think this is ideal.

**Comment by LogicMonsters** [comment]
@rrhvella Not sure if I fully got your ideas. But we do need more granular transformation events when the tree is changed (by Drag, Drop, etc)

**Comment by anil1712** [comment]
Hi,
I also needed the same feature. 
I want to revert the last node moved.
Can we have a prop  something like `onBeforeNodeMove` if we return true then only it should update the tree data and it would be better if we can have source and target both node information?

Thanks

**Comment by anil1712** [comment]
Hi @fritz-c 
Is there any progress going on regarding this feature?

**Comment by fritz-c** [comment]
@rrhvella Seems like a reasonable use case. I'll have to look into this once I have some time. 

Can you clarify what you meant about the tree data being state updated by the data prop? I didn't quite understand that. 

P.S., I didn't know anyone knew about the lazy loading capabilities. FYI the reason they're not documented is because they haven't been tested thoroughly, and I seem to recall some edge case bugs.

**Comment by rrhvella** [comment]
Hi @fritz-c 

What I mean is that the object which the tree is rendered from is stored inside the component's state (rows), and then updated whenever the treeData  prop changes, or an action is performed. 

So let's say I move a node and based on a set of criteria, I decide not to accept this change. The easiest way for me to do this would be simply not to update treeData, so that the rendered tree does not change. However the rendered tree is based on the rows state inside the tree component itself, and moveNode also updates rows. So since I am not updating the prop, the tree stays the same, with the node move operation which I wanted to cancel. Now,  the only way for me to revert the change is to update the treeData prop by making a shallow copy of my existing treeData prop, thereby causing rows to be rebuilt. 

Ideally, if I don't update the treeData prop, then the rendered tree does not get updated, giving me complete control of state changes.

Does that make sense?

I think I discovered the lazy loading capabilities because I was looking through the source code. For some reason I also thought I read about them somewhere.

**Comment by fritz-c** [comment]
@rrhvella Yes, that makes sense. I suppose I could try using a memoized tree util to generate the rows on each render without a performance hit.

**Comment by eljenso** [comment]
What is the status of this issue?

My use case is that I do not want to allow more than one root node.
I got the validation inside `onMoveNode` (with `if (!nextParentNode) {}`) but can't work out how to reset `treeData` to the state before the move.
Tried using `removeNodeAtPath` and `addNodeUnderParent` to move the node back into the original place. But after this, the UI looks broken:
![screenshot from 2018-10-09 12-53-09](https://user-images.githubusercontent.com/2805194/46664806-5847e200-cbc2-11e8-9398-fdec6f61f823.png)

**Comment by kserjey** [comment]
I think that more reliable way to add this feature is reducer property like [downshift](https://github.com/paypal/downshift#statereducer) has!

**Comment by vignesh-vimana** [comment]
Hey all, is there any clean way to do this yet? Thanks.
